src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,411 @@
+            if (normalValidator == null)
+                return checkBlocks(AIR_CHECK);
+            else {
+                return checkBlocks(normalValidator.and(AIR_CHECK));
+            }
+        }
+
+        /**
+         * Set validator for tile entities. If want to record tile entities regardless of its condition, use
+         *
+         * @return this
+         */
+        public Builder checkTiles(TriPredicate<BlockPos, IBlockState, TileEntity> tileValidator) {
+            Preconditions.checkState(!built);
+
+            this.tileValidator = tileValidator;
+            return this;
+        }
+
+        /**
+         * Record the tile entities regardless of its condition.
+         *
+         * @see #checkTiles(TriPredicate)
+         */
+        public Builder recordTiles() {
+            return checkTiles((pos, state, tile) -> true);
+        }
+
+        public RegionSnapshot build() {
+            Preconditions.checkState(!built);
+
+            ImmutableList.Builder<IBlockState> blockStates = ImmutableList.builder();
+            ImmutableList.Builder<BlockSnapshot> tileSnapshots = ImmutableList.builder();
+            if (tileValidator != null)
+                buildWithTiles(blockStates, tileSnapshots);
+            else
+                buildWithoutTiles(blockStates);
+
+            built = true;
+            return new RegionSnapshot(world, region, blockStates.build(), tileSnapshots.build());
+        }
+
+        private void buildWithTiles(ImmutableList.Builder<IBlockState> blockStates, ImmutableList.Builder<BlockSnapshot> tileSnapshots) {
+            for (BlockPos pos : region) {
+                TileEntity tile = world.getTileEntity(pos);
+                if (tile != null) {
+                    IBlockState state = world.getBlockState(pos);
+                    if (tileValidator.test(pos, state, tile))
+                        tileSnapshots.add(new BlockSnapshot(world, pos, state));
+                } else {
+                    IBlockState state = world.getBlockState(pos);
+                    if (normalValidator.test(pos, state))
+                        blockStates.add(state);
+                    else
+                        blockStates.add(DummyBlockState.INSTANCE);
+                }
+            }
+        }
+
+        private void buildWithoutTiles(ImmutableList.Builder<IBlockState> blockStates) {
+            for (BlockPos pos : region) {
+                IBlockState state = world.getBlockState(pos);
+                if (normalValidator.test(pos, state))
+                    blockStates.add(state);
+                else
+                    blockStates.add(DummyBlockState.INSTANCE);
+            }
+        }
+
+    }
+
+    /**
+     * Dummy, singleton block state class used to represent "nothing here". In other words, coordinate containing this
+     * block state will be left untouched.
+     * <p>
+     * {@code null} were not used because we cannot have {@code null} values in {@link ImmutableList}.
```

Comment:
You can have an immutable list of optional (which is how they meant it to be)...
You could just write norhing if you get to an empty optional instead of writing dumy data

Reply:
Also would someone from the outside even be able to check whether it is already predent?